sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -958,6 +958,18 @@ case class Sort(
   override protected def withNewChildInternal(newChild: LogicalPlan): Sort = copy(child = newChild)
 }
 
+/**
+ * Clustering data within the partition.
+ *
+ * @param cluster The clustering expressions
+ * @param child   Child logical plan
+ */
+case class Clustering(cluster: Seq[SortOrder], child: LogicalPlan) extends UnaryNode {
```

Comment:
shall we just use `Seq[Expression]`, for clustering, the ordering doesn't matter, we just need the clustering key.

Reply:
The `requiredOrdering` contains both cluster expressions (dynamic partition columns and bucket id) and sorts (sorting columns) :
https://github.com/apache/spark/blob/2c1c4d2614ae1ff902c244209f7ec3c79102d3e0/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/V1Writes.scala#L178-L180

I agree that ordering doesn't matter in clustering, how about expanding the `Clustering` as follows:
```scala
case class Clustering(
    cluster: Seq[Expression],
    sorts: Seq[SortOrder],
    child: LogicalPlan) extends UnaryNode {
  override def output: Seq[Attribute] = child.output
  override def outputOrdering: Seq[SortOrder] = sorts
  override protected def withNewChildInternal(newChild: LogicalPlan): Clustering =
    copy(child = newChild)
}
```
And the behavior of `Clustering` is that clustering rows by `clusters`, and in each cluster sort the data by `sorts`.

Reply:
good point, the bucketed table defines sorting columns and it's not clustering. Your proposal LGTM.